### DISCUSSION TITLE
Another Nvidia Hardware Acceleration problem

### DISCUSSION THREAD

**Discussion by trademark789** [blakeblackshear/frigate#4044] [Ask A Question]
I'm trying to get my 750ti decoding acceleration to work within the frigate docker container and I'm stuck at a problem where the cuvid codecs are not showing up in the container despite existing on the host system. 

docker-compose.yml

    version: '3.9'
    services:
      frigate:
        container_name: frigate
        privileged: true # this may not be necessary for all setups
        restart: unless-stopped
        image: blakeblackshear/frigate:stable-amd64
        devices:
          - /dev/bus/usb:/dev/bus/usb
          - /dev/vpu_service:/dev/vpu_service
          - /dev/dri/card0:/dev/dri/card0
        network_mode: "host"
        volumes:
          - /etc/localtime:/etc/localtime:ro
          - ./config/frigate.yml:/config/config.yml:ro
          - /data/frigate:/data/frigate
        runtime: nvidia
        deploy:
          resources:
            reservations:
              devices:
                - driver: nvidia
                  count: 1
                  capabilities: [gpu]
        environment:
          FRIGATE_RTSP_PASSWORD: 'xxxxxx'
          NVIDIA_VISIBLE_DEVICES: all
          NVIDIA_DRIVER_CAPABILITIES: compute,utility,video

config/frigate.yml

    mqtt:
      host: 192.168.1.xx
      port: 1883
    detectors:
      cpu1:
        type: cpu
      cpu2:
        type: cpu
    record:
      enabled: True
      expire_interval: 60
      retain:
        days: 0
        mode: all
      events:
        pre_capture: 10
        post_capture: 10
        retain:
          default: 7
          objects:
            person: 30
    cameras:
      garage1:
        ffmpeg:
          hwaccel_args: -c:v h264_cuvid
          inputs:
            - path: rtsp://frigate:{FRIGATE_RTSP_PASSWORD}@192.168.1.xx:554/cam/realmonitor?channel=1&subtype=00&authbasic
              roles:
                - rtmp
                - record
            - path: rtsp://frigate:{FRIGATE_RTSP_PASSWORD}@192.168.1.xx:554/cam/realmonitor?channel=1&subtype=01&authbasic
              roles:
                - detect
        detect:
          width: 640
          height: 400
          fps: 5
          enabled: True
        objects:
          track:
            - person
            - car

ldconfig

    ldconfig -p | grep libnv;
            libnvoptix.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvoptix.so.1
            libnvidia-vulkan-producer.so.515.65.01 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-vulkan-producer.so.515.65.01
            libnvidia-vulkan-producer.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-vulkan-producer.so
            libnvidia-tls.so.515.65.01 (libc6,x86-64, OS ABI: Linux 2.3.99) => /lib/x86_64-linux-gnu/libnvidia-tls.so.515.65.01
            libnvidia-tls.so.515.65.01 (libc6, OS ABI: Linux 2.3.99) => /lib/i386-linux-gnu/libnvidia-tls.so.515.65.01
            libnvidia-rtcore.so.515.65.01 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-rtcore.so.515.65.01
            libnvidia-ptxjitcompiler.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so.1
            libnvidia-ptxjitcompiler.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so
            libnvidia-opticalflow.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-opticalflow.so.1
            libnvidia-opticalflow.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-opticalflow.so
            libnvidia-opencl.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-opencl.so.1
            libnvidia-nvvm.so.4 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-nvvm.so.4
            libnvidia-nvvm.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-nvvm.so
            libnvidia-ngx.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-ngx.so.1
            libnvidia-ml.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-ml.so.1
            libnvidia-ml.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-ml.so
            libnvidia-gtk3.so.510.47.03 (libc6,x86-64) => /lib/libnvidia-gtk3.so.510.47.03
            libnvidia-gtk2.so.510.47.03 (libc6,x86-64) => /lib/libnvidia-gtk2.so.510.47.03
            libnvidia-glvkspirv.so.515.65.01 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-glvkspirv.so.515.65.01
            libnvidia-glvkspirv.so.515.65.01 (libc6) => /lib/i386-linux-gnu/libnvidia-glvkspirv.so.515.65.01
            libnvidia-glsi.so.515.65.01 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-glsi.so.515.65.01
            libnvidia-glsi.so.515.65.01 (libc6) => /lib/i386-linux-gnu/libnvidia-glsi.so.515.65.01
            libnvidia-glcore.so.515.65.01 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-glcore.so.515.65.01
            libnvidia-glcore.so.515.65.01 (libc6) => /lib/i386-linux-gnu/libnvidia-glcore.so.515.65.01
            libnvidia-fbc.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-fbc.so.1
            libnvidia-fbc.so.1 (libc6) => /lib/i386-linux-gnu/libnvidia-fbc.so.1
            libnvidia-fbc.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-fbc.so
            libnvidia-fbc.so (libc6) => /lib/i386-linux-gnu/libnvidia-fbc.so
            libnvidia-eglcore.so.515.65.01 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-eglcore.so.515.65.01
            libnvidia-eglcore.so.515.65.01 (libc6) => /lib/i386-linux-gnu/libnvidia-eglcore.so.515.65.01
            libnvidia-egl-wayland.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-egl-wayland.so.1
            libnvidia-egl-gbm.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-egl-gbm.so.1
            libnvidia-container.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-container.so.1
            libnvidia-container-go.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-container-go.so.1
            libnvidia-compiler.so.515.65.01 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-compiler.so.515.65.01
            libnvidia-cfg.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-cfg.so.1
            libnvidia-cfg.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-cfg.so
            libnvidia-allocator.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-allocator.so.1
            libnvidia-allocator.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-allocator.so
            libnvcuvid.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvcuvid.so.1
            libnvcuvid.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvcuvid.so

ldconfig INSIDE frigate docker container

    ldconfig -p | grep libnv
            libnvidia-ptxjitcompiler.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so.1
            libnvidia-opticalflow.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnvidia-opticalflow.so.1
            libnvidia-opencl.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.1
            libnvidia-ml.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnvidia-ml.so.1
            libnvidia-compiler.so.515.65.01 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnvidia-compiler.so.515.65.01
            libnvidia-cfg.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnvidia-cfg.so.1
            libnvidia-allocator.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnvidia-allocator.so.1
            libnvcuvid.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnvcuvid.so.1

nvidia-smi inside frigate docker container

    Recreating frigate ... done
    Attaching to frigate
    frigate    | Sat Oct  8 11:28:12 2022
    frigate    | +-----------------------------------------------------------------------------+
    frigate    | | NVIDIA-SMI 515.65.01    Driver Version: 515.65.01    CUDA Version: 11.7     |
    frigate    | |-------------------------------+----------------------+----------------------+
    frigate    | | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
    frigate    | | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
    frigate    | |                               |                      |               MIG M. |
    frigate    | |===============================+======================+======================|
    frigate    | |   0  NVIDIA GeForce ...  On   | 00000000:01:00.0 Off |                  N/A |
    frigate    | | 40%   30C    P8     1W /  38W |    241MiB /  2048MiB |      0%      Default |
    frigate    | |                               |                      |                  N/A |
    frigate    | +-------------------------------+----------------------+----------------------+
    frigate    |
    frigate    | +-----------------------------------------------------------------------------+
    frigate    | | Processes:                                                                  |
    frigate    | |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
    frigate    | |        ID   ID                                                   Usage      |
    frigate    | |=============================================================================|
    frigate    | +-----------------------------------------------------------------------------+

fffmpeg -decoders | grep cuvid on host system

    ffmpeg version 4.4.2-0ubuntu0.22.04.1 Copyright (c) 2000-2021 the FFmpeg developers
      built with gcc 11 (Ubuntu 11.2.0-19ubuntu1)
      configuration: --prefix=/usr --extra-version=0ubuntu0.22.04.1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 --enable-pocketsphinx --enable-librsvg --enable-libmfx --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libx264 --enable-shared
      libavutil      56. 70.100 / 56. 70.100
      libavcodec     58.134.100 / 58.134.100
      libavformat    58. 76.100 / 58. 76.100
      libavdevice    58. 13.100 / 58. 13.100
      libavfilter     7.110.100 /  7.110.100
      libswscale      5.  9.100 /  5.  9.100
      libswresample   3.  9.100 /  3.  9.100
      libpostproc    55.  9.100 / 55.  9.100
     V..... av1_cuvid            Nvidia CUVID AV1 decoder (codec av1)
     V..... h264_cuvid           Nvidia CUVID H264 decoder (codec h264)
     V..... hevc_cuvid           Nvidia CUVID HEVC decoder (codec hevc)
     V..... mjpeg_cuvid          Nvidia CUVID MJPEG decoder (codec mjpeg)
     V..... mpeg1_cuvid          Nvidia CUVID MPEG1VIDEO decoder (codec mpeg1video)
     V..... mpeg2_cuvid          Nvidia CUVID MPEG2VIDEO decoder (codec mpeg2video)
     V..... mpeg4_cuvid          Nvidia CUVID MPEG4 decoder (codec mpeg4)
     V..... vc1_cuvid            Nvidia CUVID VC1 decoder (codec vc1)
     V..... vp8_cuvid            Nvidia CUVID VP8 decoder (codec vp8)
     V..... vp9_cuvid            Nvidia CUVID VP9 decoder (codec vp9)

ffmpeg -decoders | grep cuvid inside docker

    ffmpeg version 4.3.1 Copyright (c) 2000-2020 the FFmpeg developers
      built with gcc 9 (Ubuntu 9.3.0-17ubuntu1~20.04)
      configuration: --disable-debug --disable-doc --disable-ffplay --enable-shared --enable-avresample --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-gpl --enable-libfreetype --enable-libvidstab --enable-libmfx --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libxcb --enable-libx265 --enable-libxvid --enable-libx264 --enable-nonfree --enable-openssl --enable-libfdk_aac --enable-postproc --enable-small --enable-version3 --enable-libzmq --extra-libs=-ldl --prefix=/opt/ffmpeg --enable-libopenjpeg --enable-libkvazaar --enable-libaom --extra-libs=-lpthread --enable-vaapi --extra-cflags=-I/opt/ffmpeg/include --extra-ldflags=-L/opt/ffmpeg/lib
      libavutil      56. 51.100 / 56. 51.100
      libavcodec     58. 91.100 / 58. 91.100
      libavformat    58. 45.100 / 58. 45.100
      libavdevice    58. 10.100 / 58. 10.100
      libavfilter     7. 85.100 /  7. 85.100
      libavresample   4.  0.  0 /  4.  0.  0
      libswscale      5.  7.100 /  5.  7.100
      libswresample   3.  7.100 /  3.  7.100
      libpostproc    55.  7.100 / 55.  7.100

Because of this problem the error I'm getting when running the container is:

    frigate    | [2022-10-08 11:39:09] watchdog.garage1               ERROR   : Ffmpeg process crashed unexpectedly for garage1.
    frigate    | [2022-10-08 11:39:09] watchdog.garage1               ERROR   : The following ffmpeg logs include the last 100 lines prior to exit.
    frigate    | [2022-10-08 11:39:09] ffmpeg.garage1.detect          ERROR   : Unknown decoder 'h264_cuvid'
    frigate    | [2022-10-08 11:39:09] ffmpeg.garage1.record_rtmp     ERROR   : Unknown decoder 'h264_cuvid'

I've hit just about every thread I can find ... any help on this is highly appreciated. Anything else I can provide let me know I'll be refreshing regularly.

edit: Another note is that if I switch acceleration to just "h264" it works but then ffmpeg doesn't show up in nvidia-smi so I don't think it's doing any hardware acceleration.

**Comment by trademark789** [comment]
realizing this should be in issues- reposting there, please close 

https://github.com/blakeblackshear/frigate/issues/4045

**Comment by NickM-27** [comment]
You're using the wrong image and you're on an older version. Change the image to `blakeblackshear/frigate:stable` and it'll work

**Reply by trademark789** [reply]
> You're using the wrong image and you're on an older version. Change the image to `blakeblackshear/frigate:stable` and it'll work

You're a hero. I wasn't even looking at the image. Teaches me for copying and pasting my original config.

**Reply by trademark789** [reply]
This is resolved. I'll post your comment on the issue as well.
